src/libmongoc/tests/test-mongoc-hello.c

```diff
@@ -0,0 +1,164 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "TestSuite.h"
+#include "test-libmongoc.h"
+#include "mongoc/mongoc.h"
+#include "mock_server/mock-server.h"
+#include "mock_server/future.h"
+#include "mock_server/future-functions.h"
+#include "mock_server/request.h"
+#include "test-conveniences.h"
+
+/* In order for these tests to pass, in addition to a successful non-terminating
+test, all of the following should /also/ be true: */
+bool
+_test_mongoc_should_check_hello ()
+{
+   return test_framework_max_wire_version_at_least (WIRE_VERSION_OP_MSG);
+}
+
+/* Example test of the first hello / isMaster sent on a mongoc_client_t.
+ * TODO: test with a mongoc_client_pool_t, which uses a separate code-path. */
+bool
+_test_mongoc_hello_impl ()
+{
+   mock_server_t *server;
+   mongoc_client_t *client;
+   bson_t *ping;
+   bson_error_t error;
+   future_t *future;
+   request_t *request;
+   bool ret;
+
+   server = mock_server_new ();
+   mock_server_run (server);
+
+   client = mongoc_client_new_from_uri (mock_server_get_uri (server));
+   ping = BCON_NEW ("ping", BCON_INT32 (1));
+
+   /* Use a "future" function to send a ping command in the background. */
+   future = future_client_command_simple (
+      client, "db", ping, NULL /* read_prefs */, NULL /* reply */, &error);
+
+   /* Since this is the first command, a new connection is opened. */
+   request = mock_server_receives_legacy_hello (server, "{'isMaster': 1}");
+   ASSERT (request);
+   mock_server_replies_simple (
+      request, "{'ok': 1, 'isWritablePrimary': true, 'maxWireVersion': 14 }");
+   request_destroy (request);
+
+   /* Now expect the ping command. */
+   request = mock_server_receives_msg (
+      server, MONGOC_MSG_NONE, tmp_bson ("{'ping': 1}"));
+   ASSERT (request);
+   mock_server_replies_simple (request,
+                               "{'ok': 1, 'isWritablePrimary': true }");
+   request_destroy (request);
+
+   ret = future_get_bool (future);
+   ASSERT (ret);
+   future_destroy (future);
+
+   bson_destroy (ping);
+   mongoc_client_destroy (client);
+   mock_server_destroy (server);
+
+   /* If we made it here, everything went just fine: */
+   return true;
+}
+
+void
+test_mongoc_hello ()
+{
+   /* Be sure the test didn't pass when it should /not/ have: */
+   ASSERT (_test_mongoc_should_check_hello () && _test_mongoc_hello_impl ());
+}
+
+bool
+_test_mongoc_hello_client_pool_impl ()
+{
+   mongoc_client_t
+      *cl3; /* the command client, which we'll push back into the pool */
+   bson_t *cmd;
+   bson_error_t err;
+   bool cmd_result;
+
+   mongoc_server_api_t *requested_server_api = NULL;
```

Comment:
IIUC the proposed behavior change is to always use `OP_MSG` if a server API version is set. 

I think `requested_server_api` will need to be set to test the new behavior.

Reply:
This has been a bit confounding: it sounds like:
* if we are requested to use a specific API version, we always use OP_MSG;
* if a specific version is not requested, we try OP_QUERY;
* if OP_MSG has been negotiated, subsequent commands use it.

This makes this a bit tricky to test, as I don't know of a way to directly see what was actually negotiated. (There is no was I see so far to get messages back and forthvrom the mock server.) I'll extend this a bit more and at least be sure the client side is trying both ways.